crates/mongodb-support/src/bson_type.rs

```diff
@@ -259,7 +278,60 @@ impl BsonScalarType {
             S::Undefined => true,
             S::DbPointer => true,
             S::Symbol => true,
+            S::UUID => true,
+        }
+    }
+
+    /// True iff we consider a to be a supertype of b.
+    ///
+    /// Note that if you add more supertypes here then it is important to also update the custom
+    /// equality check in our tests in mongodb_agent_common::query::serialization::tests. Equality
+    /// needs to be transitive over supertypes, so for example if we have,
+    ///
+    /// (Double, Int), (Decimal, Double)
+    ///
+    /// then in addition to comparing ints to doubles, and doubles to decimals, we also need to compare
+    /// decimals to ints.
+    pub fn is_supertype(a: Self, b: Self) -> bool {
+        Self::common_supertype(a, b).is_some_and(|c| c == a)
+    }
+
+    /// If there is a BSON scalar type that encompasses both a and b, return it. This does not
+    /// require a and to overlap. The returned type may be equal to a or b if one is a supertype of
+    /// the other.
+    pub fn common_supertype(a: BsonScalarType, b: BsonScalarType) -> Option<BsonScalarType> {
+        fn helper(a: BsonScalarType, b: BsonScalarType) -> Option<BsonScalarType> {
+            if a == b {
+                Some(a)
+            } else if a.is_binary() && b.is_binary() {
+                Some(S::BinData)
+            } else {
+                match (a, b) {
+                    (S::Double, S::Int) => Some(S::Double),
+                    _ => None,
+                }
+            }
         }
+        helper(a, b).or_else(|| helper(b, a))
+    }
+}
```

Comment:
`common_supertype` is more general than `is_supertype`: it can provide a common ancestor type given two types that don't overlap, or don't fully overlap with each other. This is a bit of future-proofing in case we add more binary subtype scalar types. For example if a field is found to have `UUID` and also `Vector` values we want to unify those to `BinData`.